platform/mbed_itm.h

```diff
@@ -0,0 +1,58 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#if defined(DEVICE_ITM)
+
+#include "hal/itm_api.h"
+
+class SingleWireOutput : public FileHandle {
```

Comment:
To fit the pattern of the rest of the system, this would be drivers/SingleWireOutput.h - you've put the ITM bits in the HAL, and this is the corresponding C++ wrapper. So it goes with SPI and friends.

Reply:
I was hoping to eventually put it into mbed_retarget.cpp like you did with the serial class: https://github.com/kjbracey-arm/mbed-os/blob/d7e42d1f69e5a8f06441742d910d795c8ae22c85/platform/mbed_retarget.cpp#L128-L180

so that we can enable SWO from the config system. 

My intention with this PR is to start some discussion about what would be right approach. @0xc0170 @bulislaw any comments?

Reply:
I'm slightly wary of setting up a sort of "central registry" pattern where every potential device has to be plopped in there.

In #5558 there's a similar concept - the JSON choice of "default onboard network stack". You set `nsapi.default-stack` to either "LWIP" or "NANOSTACK". There's no central logic for that, but each of the two stacks has a code snippet that looks like

    #define NANOSTACK 0x82917429
    #if MBED_CONF_NSAPI_DEFAULT_STACK == NANOSTACK
    OnboardNetworkStack &OnboardNetworkStack::get_default_instance() {
        return Nanostack::get_instance();
    }
   #endif

So only one of them will define that function.

Maybe the same concept would work here. JSON option in central place, but any target or library one can add a new selectable default without touching the core code - just invent a name.

Reply:
Anyway, regardless of the "default console" selection, this needs to be exposed as a separate standalone device manually accessible as `SingleWireOutput`.

See my 18 Dec comments on #5356, where they're concentrating on alternative console, but I'm urging them to make sure that the device is always manually useable by the app without being forced to make it stdin/stdout.

The built-in `FileHandle`s in my mbed_retarget.cpp are purely to preserve legacy behaviour. If anything, they will be moved out (@geky has suggested making `Sink` public, and discussions with @sg- have led to concensus that `RawSerial` should be a `FileHandle`, thus it can do the job of `DirectSerial`).

Reply:
No problem, I'll move/rename it to mbed-os/drivers. 

I'm fine with either invocation as well, but I think we should have it documented exactly how we expect people to do it!